## Part 1: Preprocessing

In [24]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import train_test_split
#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [25]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [26]:
# Create y_df with the Attrition and Department columns
dept_attrition_df = attrition_df.copy()

# Display a sample of y
y_df=pd.DataFrame(dept_attrition_df[["Attrition","Department"]])

y_df


Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [27]:
# Create a list of at least 10 column names to use as X data



# Create X_df using your selected columns


# Show the data types for X_df
X_df = pd.DataFrame(attrition_df[['Education', 'Age', 'DistanceFromHome', 'JobSatisfaction', 'OverTime',
       'StockOptionLevel', 'WorkLifeBalance', 'YearsAtCompany', 'YearsSinceLastPromotion',
       'NumCompaniesWorked']])

X_df


Education  Age  DistanceFromHome  JobSatisfaction OverTime  \
0             2   41                 1                4      Yes   
1             1   49                 8                2       No   
2             2   37                 2                3      Yes   
3             4   33                 3                3      Yes   
4             1   27                 2                2       No   
...         ...  ...               ...              ...      ...   
1465          2   36                23                4       No   
1466          1   39                 6                1       No   
1467          3   27                 4                2      Yes   
1468          3   49                 2                2       No   
1469          3   34                 8                3       No   

      StockOptionLevel  WorkLifeBalance  YearsAtCompany  \
0                    0                1               6   
1                    1                3              10   
2                    0                3               0   
3                    0                3               8   
4                    1                3               2   
...                ...              ...             ...   
1465                 1                3               5   
1466                 1                3               7   
1467                 1                3               6   
1468                 0                2               9   
1469                 0                4               4   

      YearsSinceLastPromotion  NumCompaniesWorked  
0                           0                   8  
1                           1                   1  
2                           0                   6  
3                           3                   1  
4                           2                   9  
...                       ...                 ...  
1465                        0                   4  
1466                        1                   4  
1467                        0                   1  
1468                        0                   2  
1469                        1                   2  

[1470 rows x 10 columns]

In [28]:
def status_change_int(xData):
    if (xData['OverTime'] == 'Yes'):
        return 1
    return 0
X_df['OverTime']=X_df.apply(status_change_int,axis=1)
X_df.dtypes

Education                  int64
Age                        int64
DistanceFromHome           int64
JobSatisfaction            int64
OverTime                   int64
StockOptionLevel           int64
WorkLifeBalance            int64
YearsAtCompany             int64
YearsSinceLastPromotion    int64
NumCompaniesWorked         int64
dtype: object

In [29]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)


In [ ]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary


In [42]:
from sklearn.preprocessing import StandardScaler
X_scaler = StandardScaler().fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
#X_test_scaled = X_scaler.transform(X_test)

In [43]:
X_train_scaled

array([[ 1.05247577,  0.35415465, -0.62306194, ...,  0.85516053,
        -0.05243217,  1.696417  ],
       [ 0.07701042,  0.02387672,  1.48197727, ...,  0.17301293,
        -0.3664553 , -0.69520543],
       [ 0.07701042,  2.11563694, -0.1277586 , ..., -0.85020847,
        -0.05243217,  0.50060579],
       ...,
       [ 0.07701042,  0.13396936,  0.11989307, ...,  1.87838194,
        -0.3664553 ,  0.10200205],
       [ 1.05247577,  1.01471051, -0.99453945, ...,  3.24267714,
        -0.68047843, -0.69520543],
       [ 0.07701042, -0.41649386, -0.99453945, ...,  0.34354983,
        -0.68047843, -0.69520543]])

In [32]:
# Create a StandardScaler


# Fit the StandardScaler to the training data


# Scale the training and testing data

from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Preprocess "Department" column (one-hot encoding)
deptartmen_encoded = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)
deptartmen_encoded.fit(np.array(y_train['Department']).reshape(-1,1))
y_department_train =deptartmen_encoded.transform(np.array(y_train['Department']).reshape(-1,1))
#y_department_test = deptartmen_encoded.transform(np.array(y_df['Department']).reshape(-1,1))
y_department_train




array([[0., 0., 1.],
       [0., 1., 0.],
       [0., 1., 0.],
       ...,
       [0., 1., 0.],
       [0., 0., 1.],
       [0., 0., 1.]])

In [33]:
dept_columns = deptartmen_encoded.get_feature_names_out(['Department'])
df_dept_encoded = pd.DataFrame(y_department_train, columns=dept_columns)
df_dept_encoded
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer


# Create the output layer

Department_Human Resources  Department_Research & Development  \
0                            0.0                                0.0   
1                            0.0                                1.0   
2                            0.0                                1.0   
3                            0.0                                1.0   
4                            0.0                                1.0   
...                          ...                                ...   
1097                         0.0                                0.0   
1098                         0.0                                1.0   
1099                         0.0                                1.0   
1100                         0.0                                0.0   
1101                         0.0                                0.0   

      Department_Sales  
0                  1.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
...                ...  
1097               1.0  
1098               0.0  
1099               0.0  
1100               1.0  
1101               1.0  

[1102 rows x 3 columns]

In [34]:
attrition_encoded = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

attrition_encoded.fit(np.array(y_train['Attrition']).reshape(-1,1))
y_attrition_train =   attrition_encoded.transform(np.array(y_train['Attrition']).reshape(-1,1))
#y_attrition_test = attrition_encoded.transform(np.array(y_test['Attrition']).reshape(-1,1))
y_attrition_train


array([[1., 0.],
       [1., 0.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

In [35]:
#color_encoder = LabelEncoder()
attrition_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)
attrition_encoder.fit_transform(np.array(y_train['Attrition']).reshape(-1, 1))

# Concatenate the encoded columns to the original DataFrame
#df_processed = pd.concat([y_df, df_dept_encoded], axis=1)

# Drop the original "quality" and "color" columns
#df_processed = df_processed.drop(['Attrition', 'Department'], axis=1)

#df_processed.head()

array([[1., 0.],
       [1., 0.],
       [1., 0.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

In [58]:
attr_columns = attrition_encoder.get_feature_names_out(['Attrition'])
df_attr_encoded = pd.DataFrame(y_attrition_train)
#df_attr_encoded=
#df_processed = pd.concat([y_df, df_attr_encoded], axis=1)
#df_attr_encoded.drop(columns=['0'])
#df_attr_encoded=df_attr_encoded.drop(columns={0}, axis=1)
#df_attr_encoded.drop(columns=[0], axis=1)
df_attr_encoded=df_attr_encoded.rename(columns={0: "Attrition_No"})
df_attr_encoded=df_attr_encoded.rename(columns={1: "Attrition_Yes"})
#df_attr_encoded=df_attr_encoded.drop(columns=[1], axis=1)
#df_processed = pd.concat([y_df, df_attr_encoded], axis=1)

In [59]:
df_attr_encoded

Attrition_No  Attrition_Yes
0              1.0            0.0
1              1.0            0.0
2              1.0            0.0
3              1.0            0.0
4              1.0            0.0
...            ...            ...
1097           1.0            0.0
1098           1.0            0.0
1099           1.0            0.0
1100           1.0            0.0
1101           1.0            0.0

[1102 rows x 2 columns]

In [ ]:
#attrition_encoded = LabelEncoder()
#y_train["attrition_encoded"] = attrition_encoded.fit_transform(y_train["Attrition"])
#y_attrition_columns = attrition_encoded.get_feature_names_out(['Attrition'])
#attrition_encoded.fit(np.array(y_train['Attrition']).reshape(-1,1))
#attrition_encoded.transform(np.array(y_train['Attrition']).reshape(-1,1))
#attrition_columns = attrition_encoded.get_feature_names_out(['Attrition'])
#attrition_columns
#y_attrition_test = attrition_encoded.transform(np.array(y_test['Attrition']).reshape(-1,1))
#df_attr_encoded = pd.DataFrame(y_attrition_train, columns=y_attrition_columns)
#df_attr_encoded
#y_attrition_columns

In [ ]:
#attrition_encoder = LabelEncoder()
#y_train['attrition_encoded'] = attrition_encoder.fit_transform(y_train['Attrition'])

#df_dept_encoded.values
# Concatenate the encoded columns to the original DataFrame
#df_processed = pd.concat([y_train, df_dept_encoded], axis=1)

# Drop the original "quality" and "color" columns
#df_processed = df_processed.drop(['Department', 'Attrition'], axis=1)

#df_processed

In [60]:
y_attrition = df_attr_encoded

y_department = df_dept_encoded
#df_processed[['Department_Human Resources', 'Department_Research & Development', 'Department_Sales']]
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer


# Create the output layer


In [61]:
y_department

Department_Human Resources  Department_Research & Development  \
0                            0.0                                0.0   
1                            0.0                                1.0   
2                            0.0                                1.0   
3                            0.0                                1.0   
4                            0.0                                1.0   
...                          ...                                ...   
1097                         0.0                                0.0   
1098                         0.0                                1.0   
1099                         0.0                                1.0   
1100                         0.0                                0.0   
1101                         0.0                                0.0   

      Department_Sales  
0                  1.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
...                ...  
1097               1.0  
1098               0.0  
1099               0.0  
1100               1.0  
1101               1.0  

[1102 rows x 3 columns]

In [62]:
y_attrition

Attrition_No  Attrition_Yes
0              1.0            0.0
1              1.0            0.0
2              1.0            0.0
3              1.0            0.0
4              1.0            0.0
...            ...            ...
1097           1.0            0.0
1098           1.0            0.0
1099           1.0            0.0
1100           1.0            0.0
1101           1.0            0.0

[1102 rows x 2 columns]

In [14]:
X_df

Education  Age  DistanceFromHome  JobSatisfaction  OverTime  \
0             2   41                 1                4         1   
1             1   49                 8                2         0   
2             2   37                 2                3         1   
3             4   33                 3                3         1   
4             1   27                 2                2         0   
...         ...  ...               ...              ...       ...   
1465          2   36                23                4         0   
1466          1   39                 6                1         0   
1467          3   27                 4                2         1   
1468          3   49                 2                2         0   
1469          3   34                 8                3         0   

      StockOptionLevel  WorkLifeBalance  YearsAtCompany  \
0                    0                1               6   
1                    1                3              10   
2                    0                3               0   
3                    0                3               8   
4                    1                3               2   
...                ...              ...             ...   
1465                 1                3               5   
1466                 1                3               7   
1467                 1                3               6   
1468                 0                2               9   
1469                 0                4               4   

      YearsSinceLastPromotion  NumCompaniesWorked  
0                           0                   8  
1                           1                   1  
2                           0                   6  
3                           3                   1  
4                           2                   9  
...                       ...                 ...  
1465                        0                   4  
1466                        1                   4  
1467                        0                   1  
1468                        0                   2  
1469                        1                   2  

[1470 rows x 10 columns]

In [63]:
from sklearn.model_selection import train_test_split
X_scale_train, X_scale_test, y_attr_train, y_attr_test, y_dept_train, y_dept_test = train_test_split(X_train_scaled, y_attrition, y_department)



In [64]:
input_layer = layers.Input(shape=(X_train_scaled.shape[1],), name='input')

# Shared hidden layers
shared_layer1 = layers.Dense(64, activation='relu', name='shared1')(input_layer)
shared_layer2 = layers.Dense(128, activation='relu', name='shared2')(shared_layer1)

In [65]:
deptartment_hidden = layers.Dense(32, activation='softmax', name='deptartment_hidden')(shared_layer2)

# Branch for color prediction
attrition_hidden = layers.Dense(32, activation='sigmoid', name='attrition_hidden')(shared_layer2)
dept_output = layers.Dense(3, activation='softmax', name='dept_output')(deptartment_hidden)

# Branch for color prediction
attrition_output = layers.Dense(2, activation='sigmoid', name='attrition_output')(attrition_hidden)



In [66]:
dept_output

<KerasTensor shape=(None, 3), dtype=float32, sparse=False, name=keras_tensor_20>

In [67]:
model = Model(inputs=input_layer, outputs=[dept_output, attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'dept_output': 'categorical_crossentropy', 'attrition_output': 'categorical_crossentropy'},
              metrics={'dept_output': 'accuracy', 'attrition_output': 'accuracy'})

# Display the model summary
model.summary()

# Create the model


# Compile the model


# Summarize the model


Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input (InputLayer)        │ (None, 10)             │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared1 (Dense)           │ (None, 64)             │            704 │ input[0][0]            │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ shared2 (Dense)           │ (None, 128)            │          8,320 │ shared1[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ deptartment_hidden        │ (None, 32)             │          4,128 │ shared2[0][0]          │
│ (Dense)                   │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_hidden (Dense)  │ (None, 32)             │          4,128 │ shared2[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dept_output (Dense)       │ (None, 3)              │             99 │ deptartment_hidden[0]… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ attrition_hidden[0][0] │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 17,445 (68.14 KB)

 Trainable params: 17,445 (68.14 KB)

 Non-trainable params: 0 (0.00 B)

In [68]:
attrition_hidden

<KerasTensor shape=(None, 32), dtype=float32, sparse=False, name=keras_tensor_19>

In [69]:
# Train the model
model.fit(
    X_scale_train,
    {'dept_output': y_dept_train, 'attrition_output': y_attr_train},
    epochs=100,
    batch_size=32,
    validation_split=0.2
)





Epoch 1/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 3s 28ms/step - attrition_output_accuracy: 0.3238 - dept_output_accuracy: 0.3132 - loss: 1.9777 - val_attrition_output_accuracy: 0.8735 - val_dept_output_accuracy: 0.6024 - val_loss: 1.4828
Epoch 2/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - attrition_output_accuracy: 0.8232 - dept_output_accuracy: 0.6715 - loss: 1.4810 - val_attrition_output_accuracy: 0.8735 - val_dept_output_accuracy: 0.6024 - val_loss: 1.3916
Epoch 3/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - attrition_output_accuracy: 0.8518 - dept_output_accuracy: 0.6798 - loss: 1.3482 - val_attrition_output_accuracy: 0.8735 - val_dept_output_accuracy: 0.6024 - val_loss: 1.3467
Epoch 4/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - attrition_output_accuracy: 0.8466 - dept_output_accuracy: 0.6615 - loss: 1.3111 - val_attrition_output_accuracy: 0.8735 - val_dept_output_accuracy: 0.6024 - val_loss: 1.3238
Epoch 5/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - attrition_output_accuracy: 0.8448 - de

In [70]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_scale_test, {'dept_output': y_dept_test, 'attrition_output': y_attr_test})
test_results

9/9 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.8349 - dept_output_accuracy: 0.6182 - loss: 1.7608 


[1.734701156616211, 0.8297101259231567, 0.6050724387168884]

# Print the accuracy for both department and attrition

#### Department Accuracy = 0.62
#### Attrition Accuracy = 0.835
#### Loss = 1.7608
#### Attrition loss = 0.83
#### Department loss = 0.605
#### Total is 1.76

In [ ]:
# Create a OneHotEncoder for the Department column


# Fit the encoder to the training data


# Create two new variables by applying the encoder
# to the training and testing data




In [ ]:
# Create a OneHotEncoder for the Attrition column


# Fit the encoder to the training data


# Create two new variables by applying the encoder
# to the training and testing data



## Create, Compile, and Train the Model

In [ ]:
# Find the number of columns in the X training data


# Create the input layer


# Create at least two shared layers


In [ ]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer


# Create the output layer



In [ ]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer


# Create the output layer



In [ ]:
# Create the model


# Compile the model


# Summarize the model


In [ ]:
# Train the model



In [ ]:
# Evaluate the model with the testing data


In [ ]:
# Print the accuracy for both department and attrition


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Yes Accuracy is good for both Attrition and Department
2. Sigma good acccuracy score
3. More data for training
